tinywl/tinywl.c

```diff
@@ -855,9 +859,11 @@ static void server_new_xdg_surface(struct wl_listener *listener, void *data) {
 	/* Set the scene_nodes decoration data */
 	toplevel->opacity = 1;
 	toplevel->corner_radius = 20;
-	toplevel->shadow_data = shadow_data_get_default();
-	toplevel->shadow_data.enabled = true;
-	toplevel->shadow_data.color = (struct wlr_render_color) {1.0f, 0.0f, 0.0f, 1.0f};
+
+	float blur_sigma = 20.0f;
+	toplevel->shadow = wlr_scene_shadow_create(&toplevel->server->scene->tree,
```

Comment:
Set this to toplevel->scene_tree so that the xdg surface and the shadow node share the same parent. By doing this you should be able to set the shadow position once (x/y: -blur_radius) ;)